In [1]:
import numpy as np
import pandas as pd

In [2]:
movies = pd.read_csv('tmdb_5000_movies.csv')
credits = pd.read_csv('tmdb_5000_credits.csv') 

In [3]:
movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [4]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [5]:
movies.shape

(4803, 20)

In [6]:
credits.shape

(4803, 4)

In [7]:
movies = movies.merge(credits,on='title')

In [8]:
movies.shape

(4809, 23)

In [9]:
movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [10]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

In [11]:
movies = movies[['movie_id','title','overview','genres','keywords','cast','crew']]

In [12]:
movies.shape

(4809, 7)

In [13]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [14]:
#check for missing values 
movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [15]:
#remove the rows with missing value becaues number of rows with missing value is less
movies.dropna(inplace=True)

In [16]:
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [17]:
#check for duplicated rows
movies.duplicated().sum() 

np.int64(0)

In [18]:
movies['genres'][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

- we need to convert each row
- **for example**:
'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]' **->** i need to convert this to ['Action','Adventure','Fantasy','Science Fiction']

In [19]:
import ast
def convert(obj):
    L = []
    for i in ast.literal_eval(obj):
        L.append(i['name'])
    return L

In [20]:
movies['genres'] = movies['genres'].apply(convert)

In [21]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [22]:
movies['keywords'][0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [23]:
movies['keywords'] = movies['keywords'].apply(convert)

In [24]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [25]:
movies['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [26]:
def convert_cast(obj):
    L = []
    counter = 0 
    for i in ast.literal_eval(obj):
        if counter != 3:
            L.append(i['name'])
            counter += 1 
        else:
            break 
    return L

In [27]:
movies['cast'] = movies['cast'].apply(convert_cast)

In [28]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]   
3      [Christian Bale, Michael Caine, Gary Oldman]   
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [29]:
movies['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [30]:
def convert_crew(obj):
    L = []
    for i in ast.literal_eval(obj):
        if i['job'] == 'Director':
            L.append(i['name'])
            break
    return L

In [31]:
movies['crew'] = movies['crew'].apply(convert_crew)

In [32]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [33]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [34]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())

In [35]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [36]:
movies['genres'] = movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])
movies['keywords'] = movies['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
movies['cast'] = movies['cast'].apply(lambda x:[i.replace(" ","") for i in x])
movies['crew'] = movies['crew'].apply(lambda x:[i.replace(" ","") for i in x])

In [37]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

In [38]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew'] 

In [39]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [40]:
new_df = movies[['movie_id','title','tags']]

In [41]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [42]:
new_df['tags'] = new_df['tags'].apply(lambda x:" ".join(x))

C:\Users\hanum\AppData\Local\Temp\ipykernel_22320\3089450492.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x:" ".join(x))


In [43]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [44]:
#convert the values in tags column to lower case 
new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())

C:\Users\hanum\AppData\Local\Temp\ipykernel_22320\3232516043.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(lambda x:x.lower())


In [45]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

In [46]:
# !pip install nltk

In [47]:
import nltk

from nltk.stem.porter import PorterStemmer

ps = PorterStemmer()

In [48]:
def stem(text):
    y = []

    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [49]:
new_df['tags'] = new_df['tags'].apply(stem)

C:\Users\hanum\AppData\Local\Temp\ipykernel_22320\3213734980.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df['tags'].apply(stem)


In [50]:
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features=5000,stop_words='english') 

In [51]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [52]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], shape=(4806, 5000))

In [53]:
vectors.shape

(4806, 5000)

In [54]:
cv.get_feature_names_out().tolist()

['000',
 '007',
 '10',
 '100',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '17th',
 '18',
 '18th',
 '18thcenturi',
 '19',
 '1910',
 '1920',
 '1930',
 '1940',
 '1944',
 '1950',
 '1950s',
 '1960',
 '1960s',
 '1970',
 '1970s',
 '1971',
 '1974',
 '1976',
 '1980',
 '1985',
 '1990',
 '1999',
 '19th',
 '19thcenturi',
 '20',
 '200',
 '2003',
 '2009',
 '20th',
 '21st',
 '23',
 '24',
 '25',
 '30',
 '300',
 '3d',
 '40',
 '50',
 '500',
 '60',
 '70',
 '80',
 'aaron',
 'aaroneckhart',
 'abandon',
 'abduct',
 'abigailbreslin',
 'abil',
 'abl',
 'aboard',
 'abov',
 'abus',
 'academ',
 'academi',
 'accept',
 'access',
 'accid',
 'accident',
 'acclaim',
 'accompani',
 'accomplish',
 'account',
 'accus',
 'ace',
 'achiev',
 'acquaint',
 'act',
 'action',
 'actionhero',
 'activ',
 'activist',
 'activities',
 'actor',
 'actress',
 'actual',
 'ad',
 'adam',
 'adamsandl',
 'adamshankman',
 'adapt',
 'add',
 'addict',
 'adjust',
 'admir',
 'admit',
 'adolesc',
 'adopt',
 'ador',
 'adrienbrodi',
 'adult'

In [55]:
from sklearn.metrics.pairwise import cosine_similarity 

In [56]:
similarity = cosine_similarity(vectors)

In [57]:
# similarity

In [58]:
# sorted(list(enumerate(similarity[0])),key=lambda x:x[1],reverse=True)[1:6]

In [59]:
# new_df[new_df['title'] == 'Avatar'].index[0]
movie_index = new_df[new_df['title'] == 'Avatar'].index[0]
distance = similarity[movie_index]
movie_list = sorted(list(enumerate(distance)),key=lambda x:x[1],reverse=True)[1:6]
movie_list

[(1214, np.float64(0.28676966733820225)),
 (2405, np.float64(0.26901379342448517)),
 (3728, np.float64(0.2605130246476754)),
 (507, np.float64(0.255608593705383)),
 (539, np.float64(0.25038669783359574))]

In [60]:
def recommend(movie):
    movie_index = new_df[new_df['title'] == movie].index[0]
    distance = similarity[movie_index]
    movie_list = sorted(list(enumerate(distance)),key=lambda x:x[1],reverse=True)[1:6]

    for i in movie_list:
        # print(new_df.iloc[i[0]]['title'])
        print(i)

In [61]:
recommend('Avatar')

(1214, np.float64(0.28676966733820225))
(2405, np.float64(0.26901379342448517))
(3728, np.float64(0.2605130246476754))
(507, np.float64(0.255608593705383))
(539, np.float64(0.25038669783359574))


In [62]:
# movies.iloc[2405]['title']

In [63]:
# import pickle

In [64]:
# pickle.dump(new_df,open('movies.pkl','wb'))

In [65]:
# pickle.dump(movies,open('df.pkl','wb'))

In [66]:
# pickle.dump(similarity,open('similarity.pkl','wb'))

In [67]:
# a = [1,2,3,4,5]
# e = enumerate(a) 

In [68]:
movies.iloc[0]

movie_id                                                19995
title                                                  Avatar
overview    [In, the, 22nd, century,, a, paraplegic, Marin...
genres           [Action, Adventure, Fantasy, ScienceFiction]
keywords    [cultureclash, future, spacewar, spacecolony, ...
cast            [SamWorthington, ZoeSaldana, SigourneyWeaver]
crew                                           [JamesCameron]
tags        [In, the, 22nd, century,, a, paraplegic, Marin...
Name: 0, dtype: object

In [69]:
" ".join(movies[movies['title']=='Avatar']['overview'][0])

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [70]:
# for i in movies[movies['title']=='Avatar']['genres'][0]:
#     print(i,end=' , ')

In [71]:
" ".join(movies[movies['title']=='The Avengers']['overview'].tolist()[0])

'When an unexpected enemy emerges and threatens global safety and security, Nick Fury, director of the international peacekeeping agency known as S.H.I.E.L.D., finds himself in need of a team to pull the world back from the brink of disaster. Spanning the globe, a daring recruitment effort begins!'

In [81]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [130]:
# import requests
# import time

# def fetch_movie_id(movie_name):
#     url = f"http://www.omdbapi.com/?t={movie_name}&apikey=3cbcd07b" 
    
#     for attempt in range(5):  # retry up to 3 times
#         try:
#             response = requests.get(url, timeout=5)   # set timeout
#             data = response.json()

#             # If API returns valid response
#             if data.get("Response") == "True":
#                 return {
#                     "imdbID": data.get("imdbID"),
#                     "imdbRating": data.get("imdbRating"),
#                     "imdbVotes": data.get("imdbVotes"),
#                     "Genre": data.get("Genre"),
#                     "Release Date": data.get("Released"),
#                     "Runtime": data.get("Runtime"),
#                     "Language": data.get("Language"),
#                 }
            
#             return None  # movie not found

#         except Exception as e:
#             print(f"Error for '{movie_name}' (attempt {attempt+1}): {e}")
#             time.sleep(1)  # wait before retry
    
#     return None  # failed all 3 attempts


In [83]:
# movies.loc[0:999, "id"] = movies.loc[0:999, "title"].apply(lambda x: (time.sleep(0.2), fetch_movie_id(x))[1]) 

Error for 'Real Steel' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Hall Pass' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'This Means War' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Meet the Parents' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Muppets Most Wanted' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)


In [96]:
# import pickle 
# pickle.dump(movies,open('mov1','wb')) 

In [98]:
movies.loc[1000:1999, "id"] = movies.loc[1000:1999, "title"].apply(lambda x: (time.sleep(0.2), fetch_movie_id(x))[1]) 

Error for 'Charlie St. Cloud' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Red Riding Hood' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Horrible Bosses 2' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The English Patient' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The English Patient' (attempt 2): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Predator 2' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)


In [101]:
# pickle.dump(movies,open('mov2','wb')) 

In [112]:
# movies.loc[2000:2999, "id"] = movies.loc[2000:2999, "title"].apply(lambda x: (time.sleep(0.2), fetch_movie_id(x))[1]) 

Error for 'Josie and the Pussycats' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Rage: Carrie 2' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Best Little Whorehouse in Texas' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'How She Move' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Apparition' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Apparition' (attempt 2): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Fur: An Imaginary Portrait of Diane Arbus' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Verdict' (attempt 1): HTTPConnectionPool(hos

In [113]:
# pickle.dump(movies,open('mov3','wb'))

In [124]:
# movies[movies['id'].isna()][:40]

movie_id                                              title  \
215       1979             Fantastic 4: Rise of the Silver Surfer   
283      18487                         The Taking of Pelham 1 2 3   
745      10040            Miss Congeniality 2: Armed and Fabulous   
848      72197      The Pirates! In an Adventure with Scientists!   
994      11863                                               Jade   
995      18501                                              Gamer   
996     109491                                Beautiful Creatures   
997       9275                                   Death to Smoochy   
998     329833                                        Zoolander 2   
999      12634                                     The Big Bounce   
1143    280391                      Why I Did (Not) Eat My Father   
1156     12279                            Spy Kids 3-D: Game Over   
1342     16911                               The Inhabited Island   
1422     35019                      Nanny McPhee and the Big Bang   
1489      9763                            Goal!: The Dream Begins   
1513     12920                  Dreamer: Inspired By a True Story   
1564     16340                        Rugrats in Paris: The Movie   
1587     36593                The Naked Gun 33⅓: The Final Insult   
1658     57089                      Hoodwinked Too! Hood VS. Evil   
1682    367961                        Savva. Heart of the Warrior   
1771     56288                Spy Kids: All the Time in the World   
1858     14317                         The Wild Thornberrys Movie   
1948     11699     The Bad Lieutenant: Port of Call - New Orleans   
1952     25353                    The True Story of Puss 'n Boots   
1997      2787                                        Pitch Black   
1998     12658                                Someone Like You...   
1999    152601                                                Her   
2139     76617                                  Texas Chainsaw 3D   
2270     14538         Three Kingdoms: Resurrection of the Dragon   
2271     13250                               Butterfly on a Wheel   
2317       496  Borat: Cultural Learnings of America for Make ...   
2415     41439                             Saw: The Final Chapter   
2592     15511     VeggieTales: The Pirates Who Don't Do Anything   
2743      9644                 National Lampoon's Loaded Weapon 1   
2774     48217                          The Widow of Saint-Pierre   
2800     14574                     The Boy in the Striped Pyjamas   
2856     28353                       Def Jam's How to Be a Player   
2950     30973               The Visual Bible: The Gospel of John   
2996     74534                     The Best Exotic Marigold Hotel   
2997     19405                               Recess: School's Out   

                                               overview  \
215   [The, Fantastic, Four, return, to, the, big, s...   
283   [Armed, men, hijack, a, New, York, City, subwa...   
745   [After, her, triumph, at, the, Miss, United, S...   
848   [The, luxuriantly, bearded, Pirate, Captain, i...   
994   [Someone, does, a, nasty, hatchet, job, on, a,...   
995   [Mind-control, technology, has, taken, society...   
996   [Ethan, Wate, just, wants, to, get, to, know, ...   
997   [Tells, the, story, of, Rainbow, Randolph,, th...   
998   [Derek, and, Hansel, are, modelling, again, wh...   
999   [A, small-time, con, artist, and, a, Hawaiian,...   
1143  [Based, on, the, novel, 'Evolution, Man', by, ...   
1156  [Carmen's, caught, in, a, virtual, reality, ga...   
1342  [On, the, threshold, of, 22nd, century,, furro...   
1422  [Nanny, McPhee, appears, at, the, door, of, a,...   
1489  [Like, millions, of, kids, around, the, world,...   
1513  [Ben, Crane, believes, that, a, severely, inju...   
1564  [The, Rugrats, are, back!, There's, Tommy, the...   
1587  [Frank, Drebin, is, persuaded, out, of, retire...   
1658  [Red, Riding, Hood, is, training, in, the, gro...

In [126]:
# movies.loc[3000:3999, "id"] = movies.loc[3000:3999, "title"].apply(lambda x: (time.sleep(0.2), fetch_movie_id(x))[1]) 

Error for 'MacGruber' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Sound of Music' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Omega Code' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Devil's Rejects' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The Lazarus Effect' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'The House of Sand' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)


In [127]:
# pickle.dump(movies,open('mov4','wb')) 

In [129]:
# print(4809-4000)
# movies['id'].isna().sum()

809


np.int64(879)

In [131]:
# movies.loc[4000:4808, "id"] = movies.loc[4000:4808, "title"].apply(lambda x: (time.sleep(0.2), fetch_movie_id(x))[1]) 

Error for 'Me and You and Everyone We Know' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)
Error for 'Penitentiary' (attempt 1): HTTPConnectionPool(host='www.omdbapi.com', port=80): Read timed out. (read timeout=5)


In [132]:
# pickle.dump(movies,open('mov5','wb')) 

In [135]:
# movies[movies['id'].isna()][:92]

movie_id                                              title  \
215       1979             Fantastic 4: Rise of the Silver Surfer   
283      18487                         The Taking of Pelham 1 2 3   
745      10040            Miss Congeniality 2: Armed and Fabulous   
848      72197      The Pirates! In an Adventure with Scientists!   
994      11863                                               Jade   
...        ...                                                ...   
4516    385636                              Juliet and Alfa Romeo   
4559    380097                         America Is Still the Place   
4562    119657                                   El Rey de Najayo   
4722     38786  The Blood of My Brother: A Story of Death in Iraq   
4738     13538                                      Straightheads   

                                               overview  \
215   [The, Fantastic, Four, return, to, the, big, s...   
283   [Armed, men, hijack, a, New, York, City, subwa...   
745   [After, her, triumph, at, the, Miss, United, S...   
848   [The, luxuriantly, bearded, Pirate, Captain, i...   
994   [Someone, does, a, nasty, hatchet, job, on, a,...   
...                                                 ...   
4516  [Tilen, (18),, an, attractive, high, school, s...   
4559  [1971, post, civil, rights, San, Francisco, se...   
4562  [The, dramatic, story, of, Julian,, a, Dominic...   
4722  [THE, BLOOD, OF, MY, BROTHER, goes, behind, th...   
4738  [There, is, instant, chemistry, between, Alice...   

                                      genres  \
215   [Adventure, Fantasy, Action, Thriller]   
283                 [Thriller, Drama, Crime]   
745                         [Action, Comedy]   
848   [Animation, Adventure, Family, Comedy]   
994     [Action, Thriller, Mystery, Romance]   
...                                      ...   
4516                         [Comedy, Drama]   
4559                                      []   
4562                          [Crime, Drama]   
4722                                      []   
4738                              [Thriller]   

                                               keywords  \
215   [fire, helicopter, surfboard, mask, satellite,...   
283   [hostage, newyorkcity, newyorksubway, subwaytr...   
745   [ransom, pressconference, ship, missamerica, f...   
848   [rivalry, stopmotion, pirate, aftercreditsstin...   
994     [callgirl, sanfrancisco, investigation, murder]   
...                                                 ...   
4516          [pornography, sex, love, teenager, death]   
4559                                                 []   
4562                                                 []   
4722                                                 []   
4738  [londonengland, countryside, rape, sex, trauma...   

                                                 cast                 crew  \
215           [IoanGruffudd, JessicaAlba, ChrisEvans]           [TimStory]   
283      [DenzelWashington, JohnTravolta, LuisGuzmán]          [TonyScott]   
745      [SandraBullock, ReginaKing, EnriqueMurciano]        [JohnPasquin]   
848          [HughGrant, BrendanGleeson, JeremyPiven]          [PeterLord]   
994   [DavidCaruso, LindaFiorentino, ChazzPalminteri]    [WilliamFriedkin]   
...                                               ...                  ...   
4516    [DarioNožićSerini, JanGerlKorenc, ŠpelaColja]       [BlažZavršnik]   
4559                                               []                   []   
4562             [MannyPérez, LuzGarcia, SergioCarlo]  [FernandoBaezMella]   
4722                                               []                   []   
4738         [GillianAnderson, DannyDyer, AdamRayner]            [DanReed]   

                                                   tags    id  
215   [The, Fantastic, Four, return, to, the, big, s...  None  
283   [Armed, men, hijack, a, New, York, City, subwa...  None  
745   [After, her, triumph, at, the, Miss, United, S.

In [137]:
# movies['id'][0]

{'imdbID': 'tt0499549',
 'imdbRating': '7.9',
 'imdbVotes': '1,448,400',
 'Genre': 'Action, Adventure, Fantasy',
 'Release Date': '18 Dec 2009',
 'Runtime': '162 min',
 'Language': 'English, Spanish'}

In [139]:
# movies['imdbID'] = movies['id'].apply(lambda x:x.get('imdbID') if x is not None else x)

In [146]:
# movies['Rating'] = movies['id'].apply(lambda x:x.get('imdbRating') if x is not None else x) 

In [148]:
# movies['votes'] = movies['id'].apply(lambda x:x.get('imdbVotes') if x is not None else x) 

In [149]:
# movies['Genre'] = movies['id'].apply(lambda x:x.get('Genre') if x is not None else x) 

In [150]:
# movies['Release_date'] = movies['id'].apply(lambda x:x.get('Release Date') if x is not None else x) 

In [151]:
# movies['Runtime'] = movies['id'].apply(lambda x:x.get('Runtime') if x is not None else x) 

In [152]:
# movies['language'] = movies['id'].apply(lambda x:x.get('Language') if x is not None else x) 

In [154]:
# movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                                  id     imdbID Rating  \
0  {'imdbID': 'tt0499549', 'imdbRating': '7.9', '...  tt0499549    7.9   
1  {'imdbID': 'tt0449088', 'imdbRating': '7.1', '...  tt0449088    7.1   
2  {'imdbID': 'tt2379713', 'imdbRating': '6.8', '...  tt2379713    6.8   
3  {'imdbID': 'tt1345836', 'imdbRating': '8.4', '...  tt1345836    8.4   
4  {'imdbID': 'tt0401729', 'imdbRating': '6.6', '...  tt0401729    6.6   

       votes                        Genre Release_date  Runtime  \
0  1,448,400   Action, Adventure, Fantasy  18 Dec 2009  162 min   
1    729,007   Action, Adventure, Fantasy  25 May 2007  169 min   
2    486,537  Action, Adventure, Thriller  06 Nov 2015  148 min   
3  1,948,580         Action, Crime, Drama  20 Jul 2012  164 min   
4    294,974    Action, Adventure, Sci-Fi  09 Mar 2012  132 min   

                                            language  
0                                   English, Spanish  
1                                            English  
2  English, German, Italian, Spanish, French, Ukr...  
3                                    English, Arabic  
4                                            English

In [167]:
# s = '134 min'
# if 'min' in s:
#     s = int(s.replace('min',''))

In [168]:
# s

134

In [176]:
# # movies['Runtime'].apply(lambda x:f"{(int(x.replace('min','')))//60} hours {(int(x.replace('min','')))%60:02d} mins" if x is not None else x)
# movies['Runtime'] = movies['Runtime'].apply(lambda x:x.replace('min','') if x is not None else x)
# movies['Runtime']

0       162 
1       169 
2       148 
3       164 
4       132 
        ... 
4804     81 
4805     95 
4806    32S 
4807    100 
4808     90 
Name: Runtime, Length: 4806, dtype: object

In [180]:
# movies['Runtime'] = movies['Runtime'].apply(lambda x:x.replace('N/A','0') if x is not None and 'N/A' in x else x)

In [184]:
# movies.loc[4806,'Runtime'] = '0'

In [185]:
# movies.loc[4806]['Runtime']

'0'

In [189]:
# movies['Runtime'] = movies['Runtime'].apply(lambda x:int(x) if x is not None else x) 

In [187]:
# movies['Runtime']

0       162.0
1       169.0
2       148.0
3       164.0
4       132.0
        ...  
4804     81.0
4805     95.0
4806      0.0
4807    100.0
4808     90.0
Name: Runtime, Length: 4806, dtype: float64

In [190]:
# movies['Runtime'] = movies['Runtime'].apply(lambda x:0 if x is None else x) 

In [197]:
# movies['Runtime'] = movies['Runtime'].apply(lambda x:f'{x//60} hours {x%60:02} mins') 

In [199]:
# movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            cast            crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]   

                                                tags  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                                  id     imdbID Rating  \
0  {'imdbID': 'tt0499549', 'imdbRating': '7.9', '...  tt0499549    7.9   

       votes                       Genre Release_date              Runtime  \
0  1,448,400  Action, Adventure, Fantasy  18 Dec 2009  2.0 hours 42.0 mins   

           language  
0  English, Spanish

In [203]:
# movies['Rating'] = movies['Rating'].apply(lambda x:x+'/'+'10' if x is not None else 'No ratings') 

In [206]:
# movies['votes'] = movies['votes'].apply(lambda x:'('+x+' votes'+')' if x is not None else '(0 votes)')

In [208]:
# movies['Rating'] = movies['Rating'] + " " + movies['votes']

In [218]:
import requests
import json
import time
import pandas as pd

BASE_URL = "https://caching.graphql.imdb.com/"
PAGE_SIZE = 25

HEADERS = {
    'accept': 'application/graphql+json, application/json',
    'content-type': 'application/json',
    'origin': 'https://www.imdb.com'
}

def get_variables(movie_id, after_cursor):
    variables = {
        "const": movie_id,
        "filter": {},
        "first": PAGE_SIZE,
        "sort": {"by": "HELPFULNESS_SCORE", "order": "DESC"}
    }
    if after_cursor:
        variables["after"] = after_cursor
    return variables


def fetch_page(movie_id, after_cursor):
    payload = {
        "query": """
        query TitleReviewsRefine($const: ID!, $filter: ReviewsFilter, $first: Int!, $sort: ReviewsSort, $after: ID) {
          title(id: $const) {
            reviews(filter: $filter, first: $first, sort: $sort, after: $after) {
              edges {
                node {
                  text {
                    originalText {
                      plainText
                    }
                  }
                }
              }
              pageInfo {
                hasNextPage
                endCursor
              }
            }
          }
        }""",
        "operationName": "TitleReviewsRefine",
        "variables": get_variables(movie_id, after_cursor)
    }

    response = requests.post(BASE_URL, headers=HEADERS, json=payload)
    response.raise_for_status()
    return response.json()


def get_all_review_text(movie_id, limit=1000):
    after_cursor = None
    texts = []

    # Each page gives 25 reviews → 1000 reviews ≈ 40 pages
    for _ in range((limit // PAGE_SIZE) + 5):  
        if len(texts) >= limit:
            break

        data = fetch_page(movie_id, after_cursor)
        edges = data["data"]["title"]["reviews"]["edges"]

        for review in edges:
            text = review["node"]["text"]["originalText"]["plainText"]
            texts.append(text)

            if len(texts) >= limit:
                break

        page_info = data["data"]["title"]["reviews"]["pageInfo"]
        if not page_info["hasNextPage"]:
            break

        after_cursor = page_info["endCursor"]
        time.sleep(0.4)

    # Return as ONE long string (your requirement)
    return " ".join(texts)



def get_reviews_dataframe(movie_id):
    combined_text = get_all_review_text(movie_id)
    return combined_text 
        
        
# movie_ids = ["tt2379713"]

# df = get_reviews_dataframe(movie_ids)

# print(df) 

In [210]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            cast            crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]   

                                                tags  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                                  id     imdbID  \
0  {'imdbID': 'tt0499549', 'imdbRating': '7.9', '...  tt0499549   

                     Rating              votes                       Genre  \
0  7.9/10 (1,448,400 votes)  (1,448,400 votes)  Action, Adventure, Fantasy   

  Release_date              Runtime          language  
0  18 Dec 2009  2.0 hours 42.0 mins  English, Spanish

In [215]:
# get_reviews_dataframe('tt0499549')

In [2]:
# movies['reviews'] = movies['imdbID'].apply(get_reviews_dataframe) 

In [3]:
import pickle 
# pickle.dump(movies,open('movies_f1.pkl','wb')) 

In [4]:
movies = pickle.load(open('movies_f1.pkl','rb')) 

In [5]:
import requests
import time
import pandas as pd
from concurrent.futures import ThreadPoolExecutor, as_completed

BASE_URL = "https://caching.graphql.imdb.com/"
PAGE_SIZE = 25
REVIEW_LIMIT = 200  # you want 200 reviews per movie

HEADERS = {
    'accept': 'application/graphql+json, application/json',
    'content-type': 'application/json',
    'origin': 'https://www.imdb.com'
}


# ---------- RETRY SAFE REQUEST ----------
def fetch_page_with_retry(payload, max_retries=6):
    for attempt in range(max_retries):
        try:
            response = requests.post(
                BASE_URL,
                headers=HEADERS,
                json=payload,
                timeout=20
            )
            response.raise_for_status()
            return response.json()

        except Exception:
            wait = 2 ** attempt  # exponential backoff
            time.sleep(wait)

    return None  # failed even after retry


# ---------- FETCH 200 REVIEWS ----------
def get_reviews(movie_id, limit=REVIEW_LIMIT):
    after_cursor = None
    texts = []
    pages = (limit // PAGE_SIZE) + 3    # around 11 pages

    for _ in range(pages):

        # payload for IMDB GraphQL
        payload = {
            "query": """
            query TitleReviewsRefine($const: ID!, $filter: ReviewsFilter, $first: Int!, $sort: ReviewsSort, $after: ID) {
              title(id: $const) {
                reviews(filter: $filter, first: $first, sort: $sort, after: $after) {
                  edges {
                    node {
                      text {
                        originalText {
                          plainText
                        }
                      }
                    }
                  }
                  pageInfo {
                    hasNextPage
                    endCursor
                  }
                }
              }
            }""",
            "operationName": "TitleReviewsRefine",
            "variables": {
                "const": movie_id,
                "filter": {},
                "first": PAGE_SIZE,
                "sort": {"by": "HELPFULNESS_SCORE", "order": "DESC"},
                **({"after": after_cursor} if after_cursor else {})
            }
        }

        data = fetch_page_with_retry(payload)
        if data is None:
            break

        edges = data["data"]["title"]["reviews"]["edges"]

        for r in edges:
            text = r["node"]["text"]["originalText"]["plainText"]
            texts.append(text)
            if len(texts) >= limit:
                return " ".join(texts)

        page = data["data"]["title"]["reviews"]["pageInfo"]

        if not page["hasNextPage"]:
            break

        after_cursor = page["endCursor"]

        time.sleep(0.25)  # small delay → avoids IMDb blocking

    return " ".join(texts)


# ---------- PARALLEL SCRAPER ----------
def fetch_for_movie(movie_id):
    try:
        return movie_id, get_reviews(movie_id)
    except:
        return movie_id, None


def scrape_all(movies_df):
    results = {}

    with ThreadPoolExecutor(max_workers=10) as executor:
        futures = {
            executor.submit(fetch_for_movie, imdb_id): imdb_id
            for imdb_id in movies_df["imdbID"]
        }

        for future in as_completed(futures):
            mid, text = future.result()
            results[mid] = text

    movies_df["reviews"] = movies_df["imdbID"].map(results)
    return movies_df


In [6]:
movies = scrape_all(movies)

In [13]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                                tags  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                                  id     imdbID  \
0  {'imdbID': 'tt0499549', 'imdbRating': '7.9', '...  tt0499549   
1  {'imdbID': 'tt0449088', 'imdbRating': '7.1', '...  tt0449088   
2  {'imdbID': 'tt2379713', 'imdbRating': '6.8', '...  tt2379713   
3  {'imdbID': 'tt1345836', 'imdbRating': '8.4', '...  tt1345836   
4  {'imdbID': 'tt0401729', 'imdbRating': '6.6', '...  tt0401729   

                     Rating              votes                        Genre  \
0  7.9/10 (1,448,400 votes)  (1,448,400 votes)   Action, Adventure, Fantasy   
1    7.1/10 (729,007 votes)    (729,007 votes)   Action, Adventure, Fantasy   
2    6.8/10 (486,537 votes)    (486,537 votes)  Action, Adventure, Thriller   
3  8.4/10 (1,948,580 votes)  (1,948,580 votes)         Action, Crime, Drama   
4    6.6/10 (294,974 votes)    (294,974 votes)    Action, Adventure, Sci-Fi   

  Release_date              Runtime  \
0  18 Dec 2009  2.0 hours 42.0 mins   
1  25 May 2007  2.0 hours 49.0 mins   
2  06 Nov 2015  2.0 hours 28.0 mins   
3  20 Jul 2012  2.0 hours 44.0 mins   
4  09 Mar 2012  2.0 hours 12.0 mins   

                                            language  \
0                                   English, Spanish   
1                                            English   
2  English, German, Italian, Spanish, French, Ukr...   
3                                    English, Arabic   
4                                            English   

                                             reviews  
0  Twelve years later, Titanic creator James Came...  
1  Pirates of the Caribbean: At World's End (2007...  
2  I have to say I was one of the few people who ...  
3  After eight years in seclusion, Batman resurfa...  
4  Of course John Carter is a long way from flawl...

In [14]:
pickle.dump(movies,open('movies_f2.pkl','wb')) 

In [2]:
# df = pickle.load(open('movies_f2.pkl','rb'))

In [17]:
# df.head()

In [19]:
import nltk
nltk.download('vader_lexicon')

from nltk.sentiment import SentimentIntensityAnalyzer
from tqdm.notebook import tqdm

sia = SentimentIntensityAnalyzer() 

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\hanum\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [22]:
movies['review_analysis'] = movies['reviews'].apply(lambda x:sia.polarity_scores(x)['compound']) 

AttributeError: 'NoneType' object has no attribute 'encode'

In [23]:
# print(sia.polarity_scores(movies['reviews'][0])['compound']) 

In [30]:
type(movies['reviews'][0])

str

In [32]:
movies['reviews'] = movies['reviews'].fillna("")

In [33]:
movies['reviews'] = movies['reviews'].astype(str) 

In [34]:
movies['reviews'].str.len().describe()

count      4806.000000
mean     197746.893050
std      110255.644985
min           0.000000
25%      100532.750000
50%      222837.500000
75%      282600.000000
max      476577.000000
Name: reviews, dtype: float64

In [1]:
import pickle 
movies = pickle.load(open('movies_f2.pkl','rb'))

In [2]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            cast            crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]   

                                                tags  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                                  id     imdbID  \
0  {'imdbID': 'tt0499549', 'imdbRating': '7.9', '...  tt0499549   

                     Rating              votes                       Genre  \
0  7.9/10 (1,448,400 votes)  (1,448,400 votes)  Action, Adventure, Fantasy   

  Release_date              Runtime          language  \
0  18 Dec 2009  2.0 hours 42.0 mins  English, Spanish   

                                             reviews  
0  Twelve years later, Titanic creator James Came...

In [3]:
# pip install torch --index-url https://download.pytorch.org/whl/cpu

In [4]:
# pip install transformers


In [5]:
# pip install sentencepiece

In [6]:
# pip install accelerate

In [9]:
from transformers import pipeline

summarizer = pipeline("summarization", model="allenai/led-base-16384")
sentiment_analyzer = pipeline("sentiment-analysis")

def summarize_in_chunks(text, chunk_size=3000):
    chunks = []
    for i in range(0, len(text), chunk_size):
        chunks.append(text[i:i+chunk_size])
    return chunks

text = movies['reviews'][0]

# Step 1: chunk summarization
chunk_summaries = []
for chunk in summarize_in_chunks(text):
    s = summarizer(chunk, max_length=300, min_length=150, do_sample=False)[0]["summary_text"]
    chunk_summaries.append(s)

combined_summary = " ".join(chunk_summaries)

# Step 2: RE-SUMMARIZE to make it < 500 tokens
final_summary = summarizer(
    combined_summary,
    max_length=400,
    min_length=200,
    do_sample=False
)[0]["summary_text"]

# Step 3: Sentiment analysis on final short summary
result = sentiment_analyzer(final_summary)[0]
print("\nSentiment:", result)


Device set to use cpu
No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cpu
Input ids are automatically padded from 604 to 1024 to be a multiple of `config.attention_window`: 1024
Both `max_new_tokens` (=256) and `max_length`(=300) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
Input ids are automatically padded from 614 to 1024 to be a multiple of `config.attention_window`: 1024
Both `max_new_tokens` (=256) and `max_length`(=300) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/tr

KeyboardInterrupt: 